obi_one/scientific/simulation/stimulus.py

```diff
@@ -349,3 +356,85 @@ def _generate_config(self) -> dict:
     spike_probability: float | list[float]
+
+
+class SpikeStimulus(Stimulus):
+    _module: str = "synapse_replay"
+    _input_type: str = "spikes"
+    duration: float
+    gid_spike_map: dict = {}
+    spike_file: Path | None = None
+    neuron_set: NeuronSetUnion
+
+    def _generate_config(self) -> dict:
+        assert self.spike_file is not None
+        sonata_config = {}
+        sonata_config[self.name] = {
+                "delay": 0.0, # If it is present, then the simulation filters out those times that are before the delay
+                "duration": self.duration,
+                "cells": self.neuron_set.name,
+                "module": self._module,
+                "input_type": self._input_type,
+                "spike_file": str(self.spike_file)
+            }
+        
+        return sonata_config
+
+    def generate_spikes(self, circuit, population, spike_file_path):
+        raise NotImplementedError("Subclasses should implement this method.")
+
+    @staticmethod
+    def write_spike_file(gid_spike_map, spike_file, neuron_set):
+        """
+        Writes SONATA output spike trains to file.
+        
+        Spike file format specs: https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#spike-file
+        """
+        # IMPORTANT: Convert SONATA node IDs (0-based) to NEURON cell IDs (1-based)!!
+        # (See https://sonata-extension.readthedocs.io/en/latest/blueconfig-projection-example.html#dat-spike-files)
+        gid_spike_map = {k + 1: v for k, v in gid_spike_map.items()}
+    
+        out_path = os.path.split(spike_file)[0]
+        if not os.path.exists(out_path):
+            os.makedirs(out_path)
+
+        popul_name = neuron_set.name
+        time_list = []
+        gid_list = []
+        for gid, spike_times in gid_spike_map.items():
+            if spike_times is not None:
+                for t in spike_times:
+                    time_list.append(t)
+                    gid_list.append(gid)
+        spike_df = pd.DataFrame(np.array([time_list, gid_list]).T, columns=['t', 'gid'])
+        spike_df = spike_df.astype({'t': float, 'gid': int})
+        spike_df.sort_values(by=['t', 'gid'], inplace=True)  # Sort by time
+        with h5py.File(spike_file, 'w') as f:
+            pop = f.create_group(f"/spikes/{popul_name}")
+            ts = pop.create_dataset("timestamps", data=spike_df['t'].values, dtype=np.float64)
+            nodes = pop.create_dataset("node_ids", data=spike_df['gid'].values, dtype=np.uint64)
+            ts.attrs['units'] = 'ms'
+
+class PoissonSpikeStimulus(SpikeStimulus):
+    _module: str = "synapse_replay"
+    _input_type: str = "spikes"
+    frequency: float  # Hz
+
+    def generate_spikes(self, circuit, population, spike_file_path):
+        gids = self.neuron_set.get_neuron_ids(circuit, population)
+        gid_spike_map = {}
+        start_time = self.timestamps.timestamps()[0]
+        end_time = start_time + self.duration
```

Comment:
At the moment, we are only using the first timestamp here (which was just a temporary solution). How about turning this into a loop iterating over all timestamps and using each of them as start of a new Poisson spike strain with the given duration?
In this way it would be possible to generate, e.g., a series of Poisson stimuli with duration XX ms every YY ms etc. It would have to be checked that the intervals are not overlapping though.
```suggestion
        timestamps = self.timestamps.timestamps()
        for t_idx, start_time in enumerate(timestamps):
            end_time = start_time + self.duration
            if t_idx < len(timestamps) - 1:
                # Check that interval not overlapping with next stimulus onset
                assert end_time < timestamps[t_idx + 1], "Stimulus time intervals overlap!"
            ...
```